skbio/stats/composition.py

```diff
@@ -2058,3 +2061,423 @@ def dirmult_ttest(
+            raise TypeError(errmsg % "Data")
+
+    if data.ndim != 2:
+        try:
+            data = pd.DataFrame(data, index=data.dtype.names)
+        except (TypeError, ValueError):
+            raise TypeError(errmsg % "Data")
+
+    if not hasattr(data, "__getitem__"):
+        try:
+            data = pd.DataFrame(data)
+        except (TypeError, ValueError):
+            raise TypeError(errmsg % "Data")
+
+    if not isinstance(metadata, pd.DataFrame):
+        try:
+            metadata = pd.DataFrame(metadata)
+        except (TypeError, ValueError):
+            raise TypeError(errmsg % "Metadata")
+
+    # Test if data has missing values
+    if data.isnull().values.any():
+        raise ValueError("Cannot handle missing values in data.")
+
+    # Test if metadata has missing values
+    if metadata.isnull().values.any():
+        raise ValueError("Cannot handle missing values in metadata.")
+
+    # Test if metadata and data have the same index, regardless of order
+    if not data.index.sort_values().equals(metadata.index.sort_values()):
+        raise ValueError("Data and metadata must have the same index.")
+
+    rng = get_rng(seed)
+
+    posterior = [
+        rng.dirichlet(data.values[i] + pseudocount) for i in range(data.shape[0])
+    ]
+
+    dir_table = pd.DataFrame(clr(posterior), index=data.index, columns=data.columns)
+
+    _call_res = _lme_call(
+        formula=formula,
+        table=dir_table,
+        metadata=metadata,
+        groups=groups,
+        reml=reml,
+        method=method,
+        fit_kwargs=fit_kwargs,
+        **kwargs,
+    )
+
+    res = _call_res[0]
+    list_of_vars = _call_res[2]
+
+    # Creating an empty dict to store sum of values (Using a DataFrame threw errors)
+    # uses a separate array for each covariate and each feature
+    # array index: 0: pvalue, 1: CI(2.5), 2: CI(97.5), 3: log2fc, 4: qvalue
+    res_dict = {}
+
+    for feature in data.columns:
+        res_ = res_dict[feature] = {}
+        for covar in list_of_vars:
+            res_[covar] = [0] * 5
+
+    for single_covar_data in res:
+        group = res_dict[single_covar_data["FeatureID"]][single_covar_data["Covariate"]]
+        for i, key in enumerate(("pvalue", "CI(2.5)", "CI(97.5)", "Log2(FC)")):
+            group[i] = single_covar_data[key]
+
+    for i in range(1, draws):
+        posterior = [
+            rng.dirichlet(data.values[i] + pseudocount) for i in range(data.shape[0])
+        ]
+        dir_table = pd.DataFrame(clr(posterior), index=data.index, columns=data.columns)
+
+        _call_ires = _lme_call(
+            formula=formula,
+            table=dir_table,
+            metadata=metadata,
+            groups=groups,
+            reml=reml,
+            method=method,
+            fit_kwargs=fit_kwargs,
+            **kwargs,
+        )
+
+        ires = _call_ires[0]
+
+        if ires is None:
+            continue
+
+        # online average to avoid holding all of the results in memory
```

Comment:
If the amount of data being tracked is well under expected system memory, like 1GB for instance, then doing the aggregate statistics after the fact rather than online could simplify the work

Reply:
Unfortunately, the amount of data being tracked could be well over system memory.

Say you have 10K samples by 100K features by 10K draws, you'd be tracking 10 trillion entries.